stetho/src/main/java/com/facebook/stetho/inspector/elements/Document.java

```diff
@@ -0,0 +1,483 @@
+  }
+
+  private ChildEventingList acquireChildEventingList(Object parentElement, DOMView domView) {
+    ChildEventingList childEventingList = mCachedChildEventingList;
+    if (childEventingList == null) {
+      childEventingList = new ChildEventingList();
+    }
+    childEventingList.acquire(parentElement, domView);
+    return childEventingList;
+  }
+
+  private void releaseChildEventingList(ChildEventingList childEventingList) {
+    childEventingList.release();
+    if (mCachedChildEventingList == null) {
+      mCachedChildEventingList = childEventingList;
+    }
+  }
+
+  private ArrayListAccumulator<Object> acquireChildrenAccumulator() {
+    ArrayListAccumulator<Object> accumulator = mCachedChildrenAccumulator;
+    if (accumulator == null) {
+      accumulator = new ArrayListAccumulator<>();
+    }
+    mCachedChildrenAccumulator = null;
+    return accumulator;
+  }
+
+  private void releaseChildrenAccumulator(ArrayListAccumulator<Object> accumulator) {
+    accumulator.clear();
+    if (mCachedChildrenAccumulator == null) {
+      mCachedChildrenAccumulator = accumulator;
+    }
+  }
+
+  private final class DOMObjectIdMapper extends ObjectIdMapper {
+    @Override
+    protected void onMapped(Object object, int id) {
+      mDOMProvider.verifyThreadAccess();
+
+      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
+      descriptor.hook(object);
+    }
+
+    @Override
+    protected void onUnmapped(Object object, int id) {
+      mDOMProvider.verifyThreadAccess();
+
+      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
+      descriptor.unhook(object);
+    }
+  }
+
+  private ShadowDOM.Update createShadowDOMUpdate() {
+    mDOMProvider.verifyThreadAccess();
+
+    if (mDOMProvider.getRootElement() != mShadowDOM.getRootElement()) {
+      throw new IllegalStateException();
+    }
+
+    ArrayListAccumulator<Object> childrenAccumulator = acquireChildrenAccumulator();
+
+    ShadowDOM.UpdateBuilder updateBuilder = mShadowDOM.beginUpdate();
+    Queue<Object> queue = new ArrayDeque<>();
+    queue.add(mDOMProvider.getRootElement());
+
+    while (!queue.isEmpty()) {
+      final Object element = queue.remove();
+      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(element);
+      mObjectIdMapper.putObject(element);
+      descriptor.getChildren(element, childrenAccumulator);
+      updateBuilder.setElementChildren(element, childrenAccumulator);
+      for (int i = 0, N = childrenAccumulator.size(); i < N; ++i) {
+        queue.add(childrenAccumulator.get(i));
+      }
+      childrenAccumulator.clear();
+    }
+
+    releaseChildrenAccumulator(childrenAccumulator);
+
+    return updateBuilder.build();
+  }
+
+  public static final class AttributeListAccumulator
+      extends ArrayList<String> implements AttributeAccumulator {
+
+    @Override
+    public void store(String name, String value) {
+      add(name);
+      add(value);
+    }
+  }
+
+  private void cleanUp() {
```

Comment:
move this lifecycle method up with addref/release so that they're all grouped together